File: bhButton/bhButton.c
#include <stdio.h>
#include <sys/time.h>
#include <wiringPi.h>

#define GPIO_UP 17
#define GPIO_DOWN 27
#define GPIO_RIGHT 22
#define GPIO_LEFT 24
#define GPIO_ENTER 25
#define BUTTON_USLEEP 50000
#define RESET_SEC 5

int setupGpio();
void isrUp();
void isrDown();
void isrRight();
void isrLeft();
void isrEnter();
void isrButton(int pin, char *button);

// 現在押されているボタン名
static volatile char *currentButton = NULL;

// 押された時間
static volatile long currentButtonTime = 0; 

// メイン関数
int main(int argc, char *argv[]) {
	struct timeval now;

	// GPIOの初期化
	if (setupGpio() == -1) return 1;

	while (1) {
		sleep(1);

		// 現在、汎用ボタンが押されている場合
		// 他の汎用ボタンを最大RESET_SEC(5秒)間無効
		if (currentButton != NULL) {
			gettimeofday(&now, NULL);
			if (RESET_SEC < now.tv_sec - currentButtonTime) {
				currentButton = NULL;
			}
		}
	}
	return 0;
}

// GPIOの初期化関数
int setupGpio() {
	// WiringPiの初期化
	if (wiringPiSetupGpio() == -1) return -1;

	// 入力モードに設定 
	pinMode(GPIO_UP, INPUT);
	pinMode(GPIO_DOWN, INPUT);
	pinMode(GPIO_RIGHT, INPUT);
	pinMode(GPIO_LEFT, INPUT);
	pinMode(GPIO_ENTER, INPUT);

	// PULLUPモードに設定 
	pullUpDnControl(GPIO_UP, PUD_UP);
	pullUpDnControl(GPIO_DOWN, PUD_UP);
	pullUpDnControl(GPIO_RIGHT, PUD_UP);
	pullUpDnControl(GPIO_LEFT, PUD_UP);
	pullUpDnControl(GPIO_ENTER, PUD_UP);

	// GPIOの値の変化を検出し、割り込み処理を行う
	wiringPiISR(GPIO_UP, INT_EDGE_BOTH, &isrUp);
	wiringPiISR(GPIO_DOWN, INT_EDGE_BOTH, &isrDown);
	wiringPiISR(GPIO_RIGHT, INT_EDGE_BOTH, &isrRight);
	wiringPiISR(GPIO_LEFT, INT_EDGE_BOTH, &isrLeft);
	wiringPiISR(GPIO_ENTER, INT_EDGE_BOTH, &isrEnter);
	return 0;
}

// upボタンの割込み関数
void isrUp() {
	isrButton(GPIO_UP, "up");
}

// downボタンの割込み関数
void isrDown() {
	isrButton(GPIO_DOWN, "down");
}

// rightボタンの割込み関数
void isrRight() {
	isrButton(GPIO_RIGHT, "right");
}

// leftボタンの割込み関数
void isrLeft() {
	isrButton(GPIO_LEFT, "left");
}

// enterボタンの割込み関数
void isrEnter() {
	isrButton(GPIO_ENTER, "enter");
}

// ボタンの割込み関数
void isrButton(int pin, char *button) {
	int read1, read2;
	struct timeval now;

	// チャタリングによる誤動作防止のため
	// BUTTON_USLEEP(0.05秒)の間隔でGPIOの値を2回取得し
	// 一致する場合のみボタン名を表示
	read1 = digitalRead(pin);
	usleep(BUTTON_USLEEP);
	read2 = digitalRead(pin);
	if (read1 == 0 && read2 == 0) {
		if (currentButton == NULL) {
			gettimeofday(&now, NULL);
			currentButtonTime = now.tv_sec;
			currentButton = button;
			printf("%s\n", button);
		}
	} else if (currentButton == button && read1 == 1 && read2 == 1) {
		currentButton = NULL;
	}
}

File: bhButton/bhButton.py
# coding: UTF-8

import RPi.GPIO as GPIO
import time

GPIO_UP = 17
GPIO_DOWN = 27
GPIO_RIGHT = 22
GPIO_LEFT = 24
GPIO_ENTER = 25
BUTTON_SLEEP = 0.05

# ボタンの割込み関数
def gpio_callback(channel):

	# チャタリングによる誤動作防止のため
	# BUTTON_SLEEP(0.05秒)の間隔でGPIOの値を2回取得し
	# 一致する場合のみボタン名を表示
	value1 = GPIO.input(channel)
	if value1 == 1:
		return
	time.sleep(BUTTON_SLEEP)
	value2 = GPIO.input(channel)
	if value2 == 1:
		return
	if value1 != value2:
		return

	if channel == GPIO_UP:
		print('up')
	elif channel == GPIO_DOWN:
		print('down')
	elif channel == GPIO_RIGHT:
		print('right')
	elif channel == GPIO_LEFT:
		print('left')
	elif channel == GPIO_ENTER:
		print('enter')

# 入力モード,PULLUPモードに設定
GPIO.setmode(GPIO.BCM)
GPIO.setup(GPIO_UP, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(GPIO_DOWN, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(GPIO_RIGHT, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(GPIO_LEFT, GPIO.IN, pull_up_down=GPIO.PUD_UP)
GPIO.setup(GPIO_ENTER, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# GPIOの値の変化(1から0)を検出し、割り込み処理を行う
GPIO.add_event_detect(GPIO_UP, GPIO.FALLING)
GPIO.add_event_detect(GPIO_DOWN, GPIO.FALLING)
GPIO.add_event_detect(GPIO_RIGHT, GPIO.FALLING)
GPIO.add_event_detect(GPIO_LEFT, GPIO.FALLING)
GPIO.add_event_detect(GPIO_ENTER, GPIO.FALLING)
GPIO.add_event_callback(GPIO_UP, gpio_callback)
GPIO.add_event_callback(GPIO_DOWN, gpio_callback)
GPIO.add_event_callback(GPIO_RIGHT, gpio_callback)
GPIO.add_event_callback(GPIO_LEFT, gpio_callback)
GPIO.add_event_callback(GPIO_ENTER, gpio_callback)

while 1:
	time.sleep(1)
